packages/firestore/src/local/indexeddb_schema_converter.ts

```diff
@@ -455,6 +464,95 @@ export class SchemaConverter implements SimpleDbSchemaConverter {
       })
       .next(() => PersistencePromise.waitFor(writes));
   }
+
+  private runOverlayMigration(
+    db: IDBDatabase,
+    transaction: SimpleDbTransaction
+  ): PersistencePromise<void> {
+    const queuesStore = transaction.store<DbMutationQueueKey, DbMutationQueue>(
+      DbMutationQueueStore
+    );
+    const mutationsStore = transaction.store<
+      DbMutationBatchKey,
+      DbMutationBatch
+    >(DbMutationBatchStore);
+
+    const promises: Array<PersistencePromise<void>> = [];
+    let userIds = new Set<string>();
+
+    return queuesStore
+      .loadAll()
+      .next(queues => {
+        for (const queue of queues) {
+          userIds = userIds.add(queue.userId);
+        }
+      })
+      .next(() => {
+        userIds.forEach(userId => {
+          const user = new User(userId);
+          const remoteDocumentCache = newIndexedDbRemoteDocumentCache(
+            this.serializer
+          );
+          const documentOverlayCache = IndexedDbDocumentOverlayCache.forUser(
+            this.serializer,
+            user
+          );
+          let allDocumentKeysForUser = documentKeySet();
+          const range = IDBKeyRange.bound(
+            [userId, BATCHID_UNKNOWN],
+            [userId, Number.POSITIVE_INFINITY]
+          );
+          promises.push(
+            mutationsStore
+              .loadAll(DbMutationBatchUserMutationsIndex, range)
+              .next(dbBatches => {
+                dbBatches.forEach(dbBatch => {
+                  hardAssert(
+                    dbBatch.userId === userId,
+                    `Cannot process batch ${dbBatch.batchId} from unexpected user`
+                  );
+                  const batch = fromDbMutationBatch(this.serializer, dbBatch);
+                  batch
+                    .keys()
+                    .forEach(
+                      key =>
+                        (allDocumentKeysForUser =
+                          allDocumentKeysForUser.add(key))
+                    );
+                });
+              })
+              .next(() => {
+                // NOTE: The index manager and the reference delegate are
+                // irrelevant for the purpose of recalculating and saving
+                // overlays. We can therefore simply use the memory
+                // implementation.
+                const memoryPersistence = new MemoryPersistence(
```

Comment:
I am surprised this does not add more to our code size, but it make sense since we always load MemoryPersistence by default.